src/platforms/atomic-kms/server/gbm_display_allocator.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gbm_display_allocator.h"
+#include "kms_framebuffer.h"
+
+#include <drm_fourcc.h>
+#include <xf86drmMode.h>
+#include <gbm.h>
+
+namespace mg = mir::graphics;
+namespace mga = mir::graphics::atomic;
+namespace geom = mir::geometry;
+
+mga::GBMDisplayAllocator::GBMDisplayAllocator(mir::Fd drm_fd, std::shared_ptr<struct gbm_device> gbm, geom::Size size)
+    : fd{std::move(drm_fd)},
+      gbm{std::move(gbm)},
+      size{size}
+{
+}
+
+auto mga::GBMDisplayAllocator::supported_formats() const -> std::vector<DRMFormat>
+{
+    // TODO: Pull out of KMS plane info
+    return { DRMFormat{DRM_FORMAT_XRGB8888}, DRMFormat{DRM_FORMAT_ARGB8888}};
+}
+
+auto mga::GBMDisplayAllocator::modifiers_for_format(DRMFormat /*format*/) const -> std::vector<uint64_t>
+{
+    // TODO: Pull out off KMS plane info
+    return {};
+}
+
+namespace
+{
+using LockedFrontBuffer = std::unique_ptr<gbm_bo, std::function<void(gbm_bo*)>>;
+
+class GBMBoFramebuffer : public mg::FBHandle
+{
+public:
+    static auto framebuffer_for_frontbuffer(mir::Fd const& drm_fd, LockedFrontBuffer bo) -> std::unique_ptr<GBMBoFramebuffer>
+    {
+        if (auto cached_fb = static_cast<std::shared_ptr<uint32_t const>*>(gbm_bo_get_user_data(bo.get())))
+        {
+            return std::unique_ptr<GBMBoFramebuffer>{new GBMBoFramebuffer{std::move(bo), *cached_fb}};
```

Comment:
AFAICS, the only reason for using raw `new` is that the constructor is private. But, as the class is local to this translation unit, and the member data is deep const, that seems a bad reason.

```suggestion
            return std::make_unique<GBMBoFramebuffer>(std::move(bo), *cached_fb);
```